CI: updates, Python 3.14, BUG: specific token values, REF: `TypeAlias`

johnyf opened the pull request:
The changes include:

- CI: test using CPython 3.12, 3.13, 3.14
- CI: install dependencies to build `cvxopt` wheel for Python 3.14
- CI: update versions of GitHub Actions used
- CI: use Ubuntu 24.04 (was 22.04)
- REF: omit `typing.TypeAlias`, which was deprecated in Python 3.12
- REL: use string as value for `project.license` and omit relevant Trove classifier, because it raises a `setuptools` error once `project.license` becomes a string value
- MNT: for editable installs use `pip install -e .` (was `python setup.py develop`)
- BUG: the lexemes `~`, `=>`, `/=`, `=<`, `<=>` propagated beyond the lexer, but instead the values `!`, `->`, `!=`, `<=`, `<->` were expected by later stages. Changed the lexer to set the value in each case to `!`, `->`, `!=`, `<=`, or `<->`.

johnyf committed:
BUG: use specific string as token value

In commit e1bae436dca2abbd72bb48a4f4b196bf3abff687,
the lexemes `~`, `=>`, `/=`, `=<`, `<=>` were added to the lexer.
These lexemes propagated as token values beyond parsing,
but the operator mappings internally expect other strings.

This change sets the token values to the expected strings.

johnyf committed:
CI: bump used actions `setup-python`, `checkout`

johnyf committed:
CI: test using CPython 3.12, 3.13, 3.14

johnyf committed:
MNT: use `pip install -e`, for `pip >= 25.3`

johnyf committed:
REL: use string value for `project.license`

This change results in `setuptools` raising an error
for the relevant Trove classifier, which is thus removed.

<https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license>
<https://peps.python.org/pep-0639/>

johnyf committed:
CI: build `cvxopt` on Python 3.14

<https://cvxopt.org/install/#ubuntu-debian>

johnyf committed:
CI: bump to Ubuntu 24.04

johnyf committed:
REF: omit `TypeAlias` (deprecated in Python 3.12)

<https://docs.python.org/3.15/library/typing.html#typing.TypeAlias>

and use the `type` statement in `contrib/xml/xmlio.py`.
<https://docs.python.org/3/reference/simple_stmts.html#type>

johnyf commented:
The tests now pass. Previous test failures were due to a dependency intermittently not installing.

slivingston commented on .github/workflows/main.yml:
```diff
@@ -13,17 +13,20 @@ on:
     build:
         name: Build
-        runs-on: ubuntu-22.04
+        runs-on: ubuntu-24.04
```
`ubuntu-22.04` is used in the `docs` job (line 71). We may want to use `ubuntu-24.04` there for consistency with this `build` job. However, it is acceptable to keep `ubuntu-22.04` for the `docs` job because Ubuntu 22.04 will be [officially supported for 18 months more](https://ubuntu.com/about/release-cycle), and I did not see a clear indication that `ubuntu-22.04` will be deprecated soon in the [runner-images repository](https://github.com/actions/runner-images).

If there is no argument against it, I recommend that we use `ubuntu-24.04` in the `docs` job, too.

johnyf replied:
I agree, I added a commit that uses `ubuntu-24.04` for the `docs` job.

slivingston commented on contrib/xml/xmlio.py:
```diff
@@ -47,7 +45,7 @@
 
 
-_Data: _ty.TypeAlias = (
+type _Data = (
```
[The `type` statement is new in Python 3.12](https://docs.python.org/3.12/whatsnew/3.12.html), so this change causes a syntax error in previous versions of Python. Because this is under contrib/, it is acceptable that the minimum version of Python to use xmlio.py is different than that of TuLiP (which is 3.10), but I wanted to comment on it to confirm that you are aware.

johnyf replied:
To avoid the error in Python 3.10 and 3.11, I added a commit that removes this `type` statement. I was considering the possibility that code under `contrib/` could require a newer Python version to run, but thinking again about it, using the `type` statement is not necessary in this case.

slivingston reviewed (approved):
Good work! I added 2 non-blocking comments.

johnyf committed:
CI: use CPython 3.14 to build documentation

johnyf committed:
MNT: use `list` for `xml.etree.ElementTree`

following API changes:
<https://docs.python.org/3.8/library/xml.etree.elementtree.html#xml.etree.ElementTree.Element.getchildren>

johnyf committed:
MNT: open XML file in binary mode

By default, `xml.etree.ElementTree.tostring()` returns
an encoded string, of type `bytes`.
<https://docs.python.org/3/library/xml.etree.elementtree.html#xml.etree.ElementTree.tostring>

johnyf committed:
BUG: typo in keyword argument to `_import_list()`

johnyf commented:
Thanks for reviewing and approving the changes. I added fixup commits that:

- to build the documentation:
  - use Ubuntu 24.04
  - use Python 3.14 (was 3.10)
  - update the `checkout` and `setup-python` GitHub Actions

  I tested these changes on branch `ci_test_docs`.

- remove the `type` statement in `xmlio.py`
- fix issues found by type checking in `xmlio.py`
